File: modules/photo-location/ios/PhotoLocationModule.swift
//
//  PhotoLocationModule.swift
//  사진 GPS 좌표만 배치로 읽는다.
//
//  왜 필요한가: expo-media-library의 getAssetInfoAsync는 좌표(asset.location)와 함께
//  localUri와 exif까지 만든다. 그 두 개가 requestContentEditingInput + CIImage(contentsOf:)로
//  원본 파일을 여는 비용이라, 실측에서 회당 15.5ms가 걸렸다(1,383회 = 21.5초, 스캔의 69.6%).
//  우리가 필요한 건 좌표뿐이고, asset.location은 Photos DB 필드라 파일을 열지 않는다.
//  fetchAssets(withLocalIdentifiers:)로 한 번에 조회하면 사실상 공짜다.
//

import ExpoModulesCore
import Photos
// CLLocationCoordinate2DIsValid를 쓰므로 명시적으로 가져온다. asset.location의 타입(CLLocation)은
// Photos를 통해 보이지만, CoreLocation의 C 함수까지 재수출된다는 보장은 없다.
import CoreLocation

public class PhotoLocationModule: Module {
    public func definition() -> ModuleDefinition {
        Name("PhotoLocation")

        // 자산 id 배열 → { id: [위도, 경도] }. 좌표가 없는 사진은 키 자체가 없다(희소 맵).
        // id는 expo-media-library가 주는 값 그대로 = PHAsset.localIdentifier.
        AsyncFunction("getLocations") { (assetIds: [String], promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                var out: [String: [Double]] = [:]
                // 좌표 있는 사진 비율이 실측 약 절반이라 절반만 잡아 둔다
                out.reserveCapacity(assetIds.count / 2 + 1)

                // 한 번의 Photos DB 조회. 없는 id는 결과에서 그냥 빠진다(에러 아님).
                let fetched = PHAsset.fetchAssets(withLocalIdentifiers: assetIds, options: nil)
                fetched.enumerateObjects { asset, _, _ in
                    if let loc = asset.location {
                        let c = loc.coordinate
                        // 좌표가 유효한지 확인 — 위치 없는 사진이 (0,0)으로 오는 경우를 막는다
                        if CLLocationCoordinate2DIsValid(c) && !(c.latitude == 0 && c.longitude == 0) {
                            out[asset.localIdentifier] = [c.latitude, c.longitude]
                        }
                    }
                }
                promise.resolve(out)
            }
        }
    }
}

File: modules/photo-location/android/src/main/java/expo/modules/photolocation/PhotoLocationModule.kt
package expo.modules.photolocation

//
//  사진 GPS 좌표만 배치로 읽는다 — iOS(PhotoLocationModule.swift)와 같은 모듈명·시그니처.
//
//  iOS와 달리 안드로이드에는 값싼 경로가 없다. 안드로이드 10(API 29)부터 스코프 스토리지가
//  MediaStore의 위치 정보를 가려서, ACCESS_MEDIA_LOCATION 권한 + setRequireOriginal로
//  원본 스트림을 열어 EXIF를 직접 읽어야 한다(LATITUDE/LONGITUDE 컬럼 경로는 사라졌다).
//  그래도 이득은 있다 — expo-media-library의 getAssetInfoAsync는 좌표 외에 localUri·크기·
//  exif 전체까지 만들고 사진 1장마다 JS↔네이티브를 왕복하는데, 여기서는 GPS 태그만 읽고
//  왕복도 배치당 1회다.
//

import android.content.ContentUris
import android.os.Build
import android.provider.MediaStore
import androidx.exifinterface.media.ExifInterface
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.Executors

class PhotoLocationModule : Module() {

    // 파일 I/O 전용 백그라운드 스레드 (JS 스레드를 막지 않는다)
    private val executor = Executors.newSingleThreadExecutor()

    override fun definition() = ModuleDefinition {
        Name("PhotoLocation")

        // 자산 id 배열 → { id: [위도, 경도] }. 좌표가 없는 사진은 키 자체가 없다(희소 맵).
        // id는 expo-media-library가 주는 값 그대로 = MediaStore의 _ID(숫자 문자열).
        AsyncFunction("getLocations") { assetIds: List<String>, promise: Promise ->
            executor.execute {
                val out = HashMap<String, List<Double>>(assetIds.size / 2 + 1)
                val resolver = appContext.reactContext?.contentResolver
                if (resolver == null) {
                    // 컨텍스트가 없으면 좌표를 못 읽는다 — 빈 맵을 주면 호출부가 기존 경로로 폴백한다
                    promise.resolve(out)
                    return@execute
                }
                for (id in assetIds) {
                    val rowId = id.toLongOrNull() ?: continue
                    try {
                        var uri = ContentUris.withAppendedId(
                            MediaStore.Images.Media.EXTERNAL_CONTENT_URI, rowId
                        )
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                            // ACCESS_MEDIA_LOCATION이 없으면 여기서 예외 → 그 사진만 건너뛴다
                            uri = MediaStore.setRequireOriginal(uri)
                        }
                        resolver.openInputStream(uri)?.use { stream ->
                            val latLong = ExifInterface(stream).latLong
                            if (latLong != null && latLong.size == 2) {
                                val lat = latLong[0]
                                val lon = latLong[1]
                                // (0,0)은 GPS 태그가 빈 값으로 박힌 사진 — 좌표 없음으로 본다
                                if (!(lat == 0.0 && lon == 0.0)) {
                                    out[id] = listOf(lat, lon)
                                }
                            }
                        }
                    } catch (e: Exception) {
                        // 권한 없음·파일 삭제됨·EXIF 없음 — 그 사진만 건너뛰고 계속한다.
                        // 한 장의 실패가 배치 전체를 무너뜨리면 안 된다.
                    }
                }
                promise.resolve(out)
            }
        }
    }
}
